From: =?utf-8?b?Ik1hcmNvIFRyZXZpc2FuIChUcmV2acOxbyki?= <mail@3v1n0.net>
Date: Mon, 1 Feb 2021 19:36:49 +0100
Subject: gdm: Expose the source serviceName for messages and verification
 failures

By giving to the AuthPrompt information regarding the source service
name (and so the ability to know whether it's a foreground service) can
give it the ability to behave differently.

Origin: https://gitlab.gnome.org/GNOME/gnome-shell/-/merge_requests/1652
Bug-Ubuntu: https://bugs.launchpad.net/ubuntu/+source/gnome-shell/+bug/1865838
---
 js/gdm/authPrompt.js |  8 ++++----
 js/gdm/util.js       | 33 +++++++++++++++++----------------
 2 files changed, 21 insertions(+), 20 deletions(-)

diff --git a/js/gdm/authPrompt.js b/js/gdm/authPrompt.js
index 083f918..bf52c75 100644
--- a/js/gdm/authPrompt.js
+++ b/js/gdm/authPrompt.js
@@ -269,12 +269,12 @@ var AuthPrompt = GObject.registerClass({
             this.reset();
     }
 
-    _onShowMessage(userVerifier, message, type) {
-        this.setMessage(message, type);
+    _onShowMessage(_userVerifier, serviceName, message, type) {
+        this.setMessage(serviceName, message, type);
         this.emit('prompted');
     }
 
-    _onVerificationFailed(userVerifier, canRetry) {
+    _onVerificationFailed(userVerifier, serviceName, canRetry) {
         this._queryingService = null;
         this.clear();
 
@@ -406,7 +406,7 @@ var AuthPrompt = GObject.registerClass({
         });
     }
 
-    setMessage(message, type) {
+    setMessage(serviceName, message, type) {
         if (type == GdmUtil.MessageType.ERROR)
             this._message.add_style_class_name('login-dialog-message-warning');
         else
diff --git a/js/gdm/util.js b/js/gdm/util.js
index d8937a4..427582d 100644
--- a/js/gdm/util.js
+++ b/js/gdm/util.js
@@ -271,7 +271,7 @@ var ShellUserVerifier = class {
 
         let message = this._messageQueue.shift();
 
-        this.emit('show-message', message.text, message.type);
+        this.emit('show-message', message.serviceName, message.text, message.type);
 
         this._messageQueueTimeoutId = GLib.timeout_add(GLib.PRIORITY_DEFAULT,
                                                        message.interval,
@@ -283,11 +283,11 @@ var ShellUserVerifier = class {
         GLib.Source.set_name_by_id(this._messageQueueTimeoutId, '[gnome-shell] this._queueMessageTimeout');
     }
 
-    _queueMessage(message, messageType) {
+    _queueMessage(serviceName, message, messageType) {
         let interval = this._getIntervalForMessage(message);
 
         this.hasPendingMessages = true;
-        this._messageQueue.push({ text: message, type: messageType, interval });
+        this._messageQueue.push({ serviceName, text: message, type: messageType, interval });
         this._queueMessageTimeout();
     }
 
@@ -298,7 +298,7 @@ var ShellUserVerifier = class {
             GLib.source_remove(this._messageQueueTimeoutId);
             this._messageQueueTimeoutId = 0;
         }
-        this.emit('show-message', null, MessageType.NONE);
+        this.emit('show-message', null, null, MessageType.NONE);
     }
 
     _checkForFingerprintReader() {
@@ -346,13 +346,13 @@ var ShellUserVerifier = class {
         }
     }
 
-    _reportInitError(where, error) {
+    _reportInitError(where, error, serviceName) {
         logError(error, where);
         this._hold.release();
 
-        this._queueMessage(_("Authentication error"), MessageType.ERROR);
+        this._queueMessage(serviceName, _('Authentication error'), MessageType.ERROR);
         this._failCounter++;
-        this._verificationFailed(false);
+        this._verificationFailed(serviceName, false);
     }
 
     _reauthenticationChannelOpened(client, result) {
@@ -469,7 +469,8 @@ var ShellUserVerifier = class {
                 } catch (e) {
                     if (e.matches(Gio.IOErrorEnum, Gio.IOErrorEnum.CANCELLED))
                         return;
-                    this._reportInitError('Failed to start %s verification'.format(serviceName), e);
+                    this._reportInitError('Failed to start %s verification'.format(serviceName), e,
+                        serviceName);
                     return;
                 }
 
@@ -487,7 +488,7 @@ var ShellUserVerifier = class {
 
     _onInfo(client, serviceName, info) {
         if (this.serviceIsForeground(serviceName)) {
-            this._queueMessage(info, MessageType.INFO);
+            this._queueMessage(serviceName, info, MessageType.INFO);
         } else if (this.serviceIsFingerprint(serviceName) && this._canRetry()) {
             // We don't show fingerprint messages directly since it's
             // not the main auth service. Instead we use the messages
@@ -495,7 +496,7 @@ var ShellUserVerifier = class {
 
             // Translators: this message is shown below the password entry field
             // to indicate the user can swipe their finger instead
-            this._queueMessage(_("(or swipe finger)"), MessageType.HINT);
+            this._queueMessage(serviceName, _('(or swipe finger)'), MessageType.HINT);
         }
     }
 
@@ -505,12 +506,12 @@ var ShellUserVerifier = class {
         if (!this.serviceIsForeground(serviceName) && !isFingerprint)
             return;
 
-        this._queueMessage(problem, MessageType.ERROR);
+        this._queueMessage(serviceName, problem, MessageType.ERROR);
         if (isFingerprint) {
             this._failCounter++;
 
             if (!this._canRetry())
-                this._verificationFailed(false);
+                this._verificationFailed(serviceName, false);
         }
     }
 
@@ -564,7 +565,7 @@ var ShellUserVerifier = class {
             (this._reauthOnly || this._failCounter < this.allowedFailures);
     }
 
-    _verificationFailed(retry) {
+    _verificationFailed(serviceName, retry) {
         // For Not Listed / enterprise logins, immediately reset
         // the dialog
         // Otherwise, when in login mode we allow ALLOWED_FAILURES attempts.
@@ -594,7 +595,7 @@ var ShellUserVerifier = class {
             }
         }
 
-        this.emit('verification-failed', canRetry);
+        this.emit('verification-failed', serviceName, canRetry);
     }
 
     _onConversationStopped(client, serviceName) {
@@ -606,7 +607,7 @@ var ShellUserVerifier = class {
         if (foregroundService) {
             this._credentialManagers[foregroundService].token = null;
             this._preemptingService = null;
-            this._verificationFailed(false);
+            this._verificationFailed(serviceName, false);
             return;
         }
 
@@ -616,7 +617,7 @@ var ShellUserVerifier = class {
         if (this.serviceIsForeground(serviceName))
             this._failCounter++;
 
-        this._verificationFailed(true);
+        this._verificationFailed(serviceName, true);
     }
 };
 Signals.addSignalMethods(ShellUserVerifier.prototype);
